source/ada/lsp-ada_handlers.adb

```diff
@@ -1438,16 +1452,259 @@ package body LSP.Ada_Handlers is
+                 Ada_Subp_Kind_Function |
+                 Ada_Subp_Kind_Procedure |
+                 Ada_Access_To_Subp_Def |
+--        Cat_Procedure
+--        Cat_Function
+--        Cat_Method
+
+                 Ada_Case_Stmt |
+--        Cat_Case_Statement
+
+                 Ada_If_Stmt |
+--        Cat_If_Statement
+
+                 Ada_For_Loop_Stmt |
+                 Ada_While_Loop_Stmt |
+--        Cat_Loop_Statement
+
+                 Ada_Begin_Block |
+                 Ada_Decl_Block |
+--        Cat_Declare_Block
+--        Cat_Simple_Block
+
+--                 Ada_Return_Stmt |
+--                 Ada_Extended_Return_Stmt |
+                 Ada_Extended_Return_Stmt_Object_Decl |
+--        Cat_Return_Block
+
+                 Ada_Select_Stmt |
+--        Cat_Select_Statement
+
+                 Ada_Entry_Body |
+--        Cat_Entry
+
+                 Ada_Exception_Handler |
+--        Cat_Exception_Handler
+
+                 Ada_Pragma_Node_List |
+                 Ada_Pragma_Argument_Assoc |
+                 Ada_Pragma_Node |
+--        Cat_Pragma
+
+                 Ada_Aspect_Assoc_List |
+                 Ada_Aspect_Assoc |
+                 Ada_Aspect_Spec =>
+--        Cat_Aspect
+
+               Store_With;
+               foldingRange.kind :=
+                 (Is_Set => True, Value => LSP.Messages.Region);
+
+               Sloc := Sloc_Range (Data (Node.Token_Start));
+               foldingRange.startLine := LSP_Number (Sloc.Start_Line) - 1;
+               if not Self.Line_Folding_Only then
+                  foldingRange.startCharacter :=
+                    (Is_Set => True,
+                     Value  => LSP_Number (Sloc.Start_Column) - 1);
+               end if;
+
+               Sloc := Sloc_Range (Data (Node.Token_End));
+               foldingRange.endLine := LSP_Number (Sloc.End_Line) - 1;
+               if not Self.Line_Folding_Only then
+                  foldingRange.endCharacter :=
+                    (Is_Set => True,
+                     Value  => LSP_Number (Sloc.End_Column) - 1);
+               end if;
+
+               if not Self.Line_Folding_Only
+                 or else foldingRange.startLine /= foldingRange.endLine
+               then
+                  Response.result.Append (foldingRange);
+               end if;
+
+            when Ada_With_Clause |
+                 Ada_Use_Package_Clause |
+                 Ada_Use_Type_Clause =>
+
+               Sloc := Sloc_Range (Data (Node.Token_Start));
+               if not Have_With then
+                  foldingRange.kind :=
+                    (Is_Set => True, Value => LSP.Messages.Imports);
+
+                  foldingRange.startLine := LSP_Number (Sloc.Start_Line) - 1;
+               end if;
+
+               foldingRange.endLine := LSP_Number (Sloc.End_Line) - 1;
+               Have_With := True;
+
+            when others =>
+               Store_With;
+         end case;
+
+         declare
```

Comment:
A proposal: use standard Libadalang traverse procedure to iterate tree nodes, then do a separate pass over trivias-comments instead of inventing a custom traverse code.

Feel free to ignore.